warehouse/accounts/views.py

```diff
@@ -169,3 +169,42 @@ def logout(request, redirect_field_name=REDIRECT_FIELD_NAME):
         return HTTPSeeOther(redirect_to, headers=dict(headers))
 
     return {"redirect": {"field": REDIRECT_FIELD_NAME, "data": redirect_to}}
+
+
+@view_config(
+    route_name="accounts.register",
+    renderer="accounts/register.html",
+    decorator=[csrf_protect("accounts.register"), uses_session],
+)
+def register(request, _form_class=forms.RegistrationForm):
+    if request.authenticated_userid is not None:
+        return HTTPSeeOther("/")
+
+    user_service = request.find_service(IUserService, context=None)
+    recaptcha_service = request.find_service(name="recaptcha")
+    recaptcha_service.add_to_csp_policy()
+
+    # the form contains an auto-generated field from recaptcha with
+    # hyphens in it. make it play nice with wtforms.
+    post_body = {
+        key.replace("-", "_"): value
+        for key, value in request.POST.items()
+    }
+
+    form = _form_class(
+        data=post_body, user_service=user_service,
+        recaptcha_service=recaptcha_service
+    )
+
+    if request.method == "POST" and form.validate():
+        user = user_service.create_user(
+            form.username.data, form.full_name.data, form.password.data,
+            form.email.data
+        )
+
+        headers = remember(request, user.id)
+
+        request.session.new_csrf_token()
```

Comment:
We should probably extract the session invalidation/csrf token/remember() call out of the login view and just call that function both from the login view and this view. Just a module scoped function in ``warehouse/accountsviews.py`` with:

```python
def _login_user(request, user_id):
    # We need to protect against session fixation attacks, so make sure
    # that we create a new session (which will cause it to get a new
    # session identifier).
    if (request.unauthenticated_userid is not None and
            request.unauthenticated_userid != user_id):
        # There is already a userid associated with this request and it is
        # a different userid than the one we're trying to remember now. In
        # this case we want to drop the existing session completely because
        # we don't want to leak any data between authenticated userids.
        request.session.invalidate()
    else:
        # We either do not have an associated userid with this request
        # already, or the userid is the same one we're trying to remember
        # now. In either case we want to keep all of the data but we want
        # to make sure that we create a new session since we're crossing
        # a privilege boundary.
        data = dict(request.session.items())
        request.session.invalidate()
        request.session.update(data)

    # Remember the userid using the authentication policy.
    headers = remember(request, user_id)

    # Cycle the CSRF token since we've crossed an authentication boundary
    # and we don't want to continue using the old one.
    request.session.new_csrf_token()

    return headers
```

should be fine I think.